### PR TITLE
MULE-9856 Create MuleMessage builder

### PR DESCRIPTION
**Pull request by dfeist**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java [hunk 1]
```diff
@@ -229,8 +229,8 @@ public void maintainsDataTypeSubClassTypeTransformation() throws Exception
 
         MuleEvent muleEvent = mock(MuleEvent.class);
 
-        MuleMessage result = transformationService.applyTransformers(muleMessage, muleEvent,
-                                                                     Collections.singletonList(transformer));
+        MuleMessage<Object, Serializable> result = transformationService.applyTransformers(muleMessage, muleEvent,
+                                                                                           Collections.singletonList(transformer));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java]: static import

File: core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java [hunk 2]
```diff
@@ -250,7 +250,7 @@ public void doesNotmaintainDataTypeDueToNotAssignableTransformation() throws Exc
 
         MuleEvent muleEvent = mock(MuleEvent.class);
 
-        MuleMessage result = transformationService.applyTransformers(muleMessage, muleEvent,
+        MuleMessage<Object, Serializable> result = transformationService.applyTransformers(muleMessage, muleEvent,
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java]: static import + format

File: core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    private static final Serializable PROPERTY_VALUE = "propertyValue";
+    private static final Object REPLY_TO = new Orange();
+
+    @Test
+    public void createNewAPIMessageViaMessageInterface()
+    {
+        org.mule.runtime.api.message.MuleMessage<String, Apple> message;
+        message = org.mule.runtime.api.message.MuleMessage.builder().payload(TEST_PAYLOAD).attributes(TEST_ATTR)
+                .build();
+        assertThat(message, instanceOf(DefaultMuleMessage.class));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java]: is this instanceof assertion really necessary? It's not like the builder may return different types.

File: core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        MuleMessage<Boolean, Apple> copy = new DefaultMuleMessageBuilder(createTestMessage()).payload(dataHandler).build();
+
+        assertThat(copy.getPayload(), is(dataHandler));
+        assertThat(copy.getDataType().getType(), equalTo(DataHandler.class));
+        assertThat(copy.getDataType().getMediaType(), is(XML));
+    }
+
+    private MuleMessage<String, Apple> createTestMessage()
+    {
+        return new DefaultMuleMessageBuilder().payload(TEST_PAYLOAD).dataType(TEXT_STRING).attributes(TEST_ATTR).rootId
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java]: make one call in a different line, for readability

File: core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    {
+        MuleMessage message = new DefaultMuleMessageBuilder().payload(TEST_PAYLOAD).addOutboundProperty(PROPERTY_KEY, PROPERTY_VALUE, HTML_STRING).build();
+        assertThat(message.getOutboundProperty(PROPERTY_KEY), equalTo(PROPERTY_VALUE));
+        assertThat(message.getOutboundPropertyDataType(PROPERTY_KEY), equalTo(HTML_STRING));
+        assertThat(message.getOutboundPropertyNames().size(), equalTo(3));
+        assertThat(message.getOutboundPropertyNames(), hasItem(PROPERTY_KEY));
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void nullPayload()
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java]: Use ExpectedException rule instead

File: core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    {
+        DefaultMuleMessageBuilder builder = new DefaultMuleMessageBuilder();
+        builder.payload(null);
+    }
+
+    @Test
+    public void mutatePayloadSameTypeConserveTypeAndMimeType()
+    {
+        MuleMessage<String, Apple> message = createTestMessage();
+        MuleMessage<Boolean, Apple> copy = new DefaultMuleMessageBuilder(message).payload("NEW_PAYLOAD").build();
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java]: extract to a constant

File: core/src/main/java/org/mule/runtime/core/DefaultMuleMessage.java [hunk 7]
```diff
@@ -95,7 +94,7 @@
     /**
      * Scoped properties for this message
      */
-    private MessagePropertiesContext properties = new MessagePropertiesContext();
+    MessagePropertiesContext properties = new MessagePropertiesContext();
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/DefaultMuleMessage.java]: need to increase visibility?

File: core/src/main/java/org/mule/runtime/core/DefaultMuleMessage.java [hunk 8]
```diff
@@ -114,8 +113,31 @@
     private transient AtomicReference<Thread> ownerThread = null;
     private transient AtomicBoolean mutable = null;
 
+    private transient TypedValue typedValue;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/DefaultMuleMessage.java]: why transient?

File: core/src/main/java/org/mule/runtime/core/api/MuleMessage.java [hunk 9]
```diff
@@ -84,21 +117,138 @@
     ExceptionPayload getExceptionPayload();
 
     /**
-     * @deprecated
-     * Avoid getting access to the MuleContext through the message.
-     * You can get access to the MuleContext by making your class implement {@link org.mule.runtime.core.api.context.MuleContextAware}
+     * @deprecated Avoid getting access to the MuleContext through the message.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/MuleMessage.java]: format

File: core/src/main/java/org/mule/runtime/core/enricher/MessageEnricher.java [hunk 10]
```diff
@@ -98,9 +100,14 @@ protected void enrich(MuleEvent currentEvent,
         else
         {
             final TypedValue finalTypedValue = typedValue;
-            currentEvent.setMessage(currentEvent.getMessage().transform(msg -> {
-                msg.setPayload(finalTypedValue.getValue(), finalTypedValue.getDataType());
-                return msg;
+            currentEvent.setMessage(currentEvent.getMessage().transform(new Function<MutableMuleMessage, MuleMessage>()
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/enricher/MessageEnricher.java]: why this change?

File: core/src/main/java/org/mule/runtime/core/DefaultMuleMessage.java (hunk 8 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/DefaultMuleMessage.java]: Because payload isn't automatically serializable.

File: core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+    @Test
+    public void messageAttributes()
+    {
+        assertTestMessage(createTestMessage());
+    }
+
+    @Test
+    public void messageAttributesCopy()
+    {
+        assertTestMessage(new DefaultMuleMessageBuilder(createTestMessage()).build());
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java]: would we prefer this method for getting this builder, instead of MuleMessage.builder() and then casting?
